### ISSUE TITLE
Database Migration is not rolled back if insert into migration table failed

### ISSUE THREAD

**Issue by jansepke** [concourse/concourse#6695]
## Summary

During the upgrade to concourse V7 our concourse web server died so that the database migration was successfully executed on the database but the entry in the migration table that the migration was completed could not be added. So the next startup of the webserver tried to execute this migration again and failed because the database state was not as expected.

## Steps to reproduce

1. start upgrade concourse to V7
2. kill web server
3. start web server again

## Expected results
Database migration will finish successfully

## Actual results
error: failed to connect to database: migration '1606068653_build_events_partitions_bigint.up.sql' failed and was rolled back: pq: index "pipeline_build_events_82_build_id" does not exist

## Additional context
We know that it would not be a problem when our web server would not get killed by the AWS ASG and we now increase the some timeout values.

Mabye wrapping the migration script and the migration table handling in one transaction could solve this.

## Triaging info

* Concourse version: 7.1.0

**Comment by Berber31** [comment]
Same issue encountered.

**Upgrade from 6.6.0 to 6.7.3** with helm chart v15.0.0:

Init container `concourse-migration` in `CrashLoopBackOff`

With this log into `concourse-migration` container:
`error: migration '1606068653_build_events_partitions_bigint.up.sql' failed and was rolled back: pq: index "pipeline_build_events_14_build_id" does not exist`

**Try to comme back to 6.6.0 failed**:
concourse web pod: 
- ready and started
- UI: experiencing turbulence
- With these logs in concourse web
`"level":"error","source":"atc","message":"atc.checker.failed-to-fetch-resource-checks","data":{"error":"pq:  │
│ relation \"checks\" does not exist","session":"20"}} `
`"level":"error","source":"atc","message":"atc.checker.tick.component-failed","data":{"error":"pq: relation \ │
│ "checks\" does not exist","session":"23.15"}}  `
`"level":"error","source":"atc","message":"atc.scanner.tick.component-failed","data":{"error":"pq: column r.c │
│ heck_error does not exist","session":"22.15"}}`

**Comment by xtremerui** [comment]
@Berber31 your case is a bit different since it was the migration failed to crash the web container versus the web server was killed by external.

So in your case if the migration failed the DB would stuck in the middle which requires you to run down migration to fully roll back the DB to the state before upgrading. Have you tried https://concourse-ci.org/concourse-web.html#downgrading ? I think once you do that it should recreate the `checks` table and you should be able to start 6.6.0 without issue.

**Comment by xtremerui** [comment]
@jansepke could you share more about the recovery of the issue? How did you fix the DB after the second time that you started the web server? I know one thing you did was increasing the timeout. Anything else? Thx a lot.

**Comment by jansepke** [comment]
there was no concourse error when it was shutdown for the first time, but we also set the concourse log level to error, so maybe we suppressed something (CONCOURSE_LOG_LEVEL=error CONCOURSE_TSA_LOG_LEVEL=error)

seems like the concourse process was killed by systemd, this is our systemd file:
```
[Unit]
Description=Concourse CI Web

[Service]
EnvironmentFile=/opt/concourse/concourse_web_env_file
ExecStart=/usr/local/concourse/bin/concourse web
User=root
Group=root
Type=simple

[Install]
WantedBy=default.target
```

So as far as I can remember we first tried to do a manual migration rollback with `concourse migrate` but that did not change anything because the migration was not added to the migration table. we then executed parts of the down script (https://github.com/concourse/concourse/blob/8a396ef118f4a7f5f39f6cb3b2442e3495ab1605/atc/db/migration/migrations/1606068653_build_events_partitions_bigint.down.sql) manually on the database until the pipeline tables looked like the up migration should work. Finally we run `concourse migrate` to execute the up migration and startet concourse successfully.

We increased the timeout later to prevent such a thing in the future
